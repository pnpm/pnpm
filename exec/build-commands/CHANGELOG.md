# @pnpm/exec.build-commands

## 1001.0.19

### Patch Changes

- @pnpm/plugin-commands-rebuild@1002.0.19

## 1001.0.18

### Patch Changes

- Updated dependencies [623da6f]
- Updated dependencies [cf630a8]
- Updated dependencies [ab155a5]
  - @pnpm/config@1004.1.0
  - @pnpm/plugin-commands-rebuild@1002.0.18
  - @pnpm/config.config-writer@1000.0.7

## 1001.0.17

### Patch Changes

- @pnpm/plugin-commands-rebuild@1002.0.17

## 1001.0.16

### Patch Changes

- Updated dependencies [b217bbb]
- Updated dependencies [b0ead51]
- Updated dependencies [c8341cc]
- Updated dependencies [b0ead51]
- Updated dependencies [046af72]
  - @pnpm/config@1004.0.0
  - @pnpm/plugin-commands-rebuild@1002.0.16
  - @pnpm/config.config-writer@1000.0.6

## 1001.0.15

### Patch Changes

- Updated dependencies [8d175c0]
  - @pnpm/config@1003.1.1
  - @pnpm/plugin-commands-rebuild@1002.0.15

## 1001.0.14

### Patch Changes

- 09cf46f: Update `@pnpm/logger` in peer dependencies.
- c00360b: Update `@pnpm/util.lex-comparator` to v3.0.2.
- Updated dependencies [b282bd1]
- Updated dependencies [fdb1d98]
- Updated dependencies [e4af08c]
- Updated dependencies [09cf46f]
- Updated dependencies [36d1448]
- Updated dependencies [9362b5f]
- Updated dependencies [6cf010c]
  - @pnpm/config@1003.1.0
  - @pnpm/plugin-commands-rebuild@1002.0.14
  - @pnpm/config.config-writer@1000.0.5
  - @pnpm/modules-yaml@1000.3.3

## 1001.0.13

### Patch Changes

- @pnpm/plugin-commands-rebuild@1002.0.13
- @pnpm/config.config-writer@1000.0.4
- @pnpm/config@1003.0.1

## 1001.0.12

### Patch Changes

- Updated dependencies [56bb69b]
- Updated dependencies [8a9f3a4]
- Updated dependencies [9c3dd03]
  - @pnpm/config@1003.0.0
  - @pnpm/logger@1001.0.0
  - @pnpm/plugin-commands-rebuild@1002.0.12
  - @pnpm/config.config-writer@1000.0.3
  - @pnpm/modules-yaml@1000.3.2

## 1001.0.11

### Patch Changes

- Updated dependencies [17b7e9f]
  - @pnpm/config.config-writer@1000.0.2
  - @pnpm/plugin-commands-rebuild@1002.0.11
  - @pnpm/config@1002.7.2

## 1001.0.10

### Patch Changes

- Updated dependencies [5679712]
- Updated dependencies [01f2bcf]
  - @pnpm/config@1002.7.1
  - @pnpm/config.config-writer@1000.0.1
  - @pnpm/plugin-commands-rebuild@1002.0.10
  - @pnpm/modules-yaml@1000.3.1

## 1001.0.9

### Patch Changes

- Updated dependencies [e57f1df]
  - @pnpm/config@1002.7.0
  - @pnpm/plugin-commands-rebuild@1002.0.9

## 1001.0.8

### Patch Changes

- Updated dependencies [9bcca9f]
- Updated dependencies [5b35dff]
- Updated dependencies [a2903a0]
- Updated dependencies [9bcca9f]
- Updated dependencies [5f7be64]
- Updated dependencies [5a9e34f]
- Updated dependencies [64f6b4f]
- Updated dependencies [5f7be64]
  - @pnpm/config@1002.6.0
  - @pnpm/plugin-commands-rebuild@1002.0.8
  - @pnpm/config.config-writer@1000.0.0
  - @pnpm/modules-yaml@1000.3.0

## 1001.0.7

### Patch Changes

- d612dcf: Remove warnings after having explicitly approved no builds [#9296](https://github.com/pnpm/pnpm/issues/9296).
- Updated dependencies [d612dcf]
- Updated dependencies [936430a]
- Updated dependencies [d612dcf]
  - @pnpm/modules-yaml@1000.2.0
  - @pnpm/config@1002.5.4
  - @pnpm/plugin-commands-rebuild@1002.0.7

## 1001.0.6

### Patch Changes

- @pnpm/plugin-commands-rebuild@1002.0.6

## 1001.0.5

### Patch Changes

- Updated dependencies [6e4459c]
  - @pnpm/config@1002.5.3
  - @pnpm/plugin-commands-rebuild@1002.0.5

## 1001.0.4

### Patch Changes

- @pnpm/plugin-commands-rebuild@1002.0.4
- @pnpm/config@1002.5.2

## 1001.0.3

### Patch Changes

- 1e6ae3e: When executing the `approve-builds` command, if package.json contains `onlyBuiltDependencies` or `ignoredBuiltDependencies`, the selected dependency package will continue to be written into `package.json`.
- Updated dependencies [c3aa4d8]
  - @pnpm/config@1002.5.1
  - @pnpm/plugin-commands-rebuild@1002.0.3

## 1001.0.2

### Patch Changes

- 8b3cfe2: fix: don't abort approve-builds command or err when manifest doesn't exist [#9198](https://github.com/pnpm/pnpm/pull/9198)
- Updated dependencies [d965748]
  - @pnpm/config@1002.5.0
  - @pnpm/plugin-commands-rebuild@1002.0.2
  - @pnpm/modules-yaml@1000.1.4
  - @pnpm/read-project-manifest@1000.0.7
  - @pnpm/workspace.manifest-writer@1000.0.2

## 1001.0.1

### Patch Changes

- Updated dependencies [23754c7]
- Updated dependencies [1c2eb8c]
  - @pnpm/workspace.manifest-writer@1000.0.1
  - @pnpm/config@1002.4.1
  - @pnpm/plugin-commands-rebuild@1002.0.1

## 1001.0.0

### Major Changes

- 8fcc221: Read `onlyBuiltDependencies` and `ignoredBuiltDependencies` from `options`.

### Patch Changes

- Updated dependencies [8fcc221]
- Updated dependencies [8fcc221]
- Updated dependencies [8fcc221]
- Updated dependencies [e32b1a2]
- Updated dependencies [8fcc221]
  - @pnpm/workspace.manifest-writer@1000.0.0
  - @pnpm/config@1002.4.0
  - @pnpm/plugin-commands-rebuild@1002.0.0
  - @pnpm/modules-yaml@1000.1.3
  - @pnpm/read-project-manifest@1000.0.6

## 1000.1.1

### Patch Changes

- 546ab37: Throws an error when the value provided by the `--allow-build` option overlaps with the `pnpm.ignoredBuildDependencies` list [#9105](https://github.com/pnpm/pnpm/pull/9105).
- Updated dependencies [fee898f]
  - @pnpm/config@1002.3.1
  - @pnpm/plugin-commands-rebuild@1001.1.8

## 1000.1.0

### Minor Changes

- 4aa6d45: `pnpm approve-builds --global` works now for allowing dependencies of globally installed packages to run postinstall scripts.

### Patch Changes

- @pnpm/plugin-commands-rebuild@1001.1.7

## 1000.0.3

### Patch Changes

- Updated dependencies [f6006f2]
  - @pnpm/config@1002.3.0
  - @pnpm/plugin-commands-rebuild@1001.1.6

## 1000.0.2

### Patch Changes

- afbb654: `pnpm approve-builds` should work, when executed from a subdirectory of a workspace [#9042](https://github.com/pnpm/pnpm/issues/9042).

## 1000.0.1

### Patch Changes

- 5d7192c: `approve-builds` command gets the auto-ignore build list and exits early when it is an empty array [#9024](https://github.com/pnpm/pnpm/pull/9024).
- a2a4509: Sort the package names in the "pnpm.onlyBuiltDependencies" list saved by `pnpm approve-builds`.
- Updated dependencies [1e229d7]
  - @pnpm/read-project-manifest@1000.0.5
  - @pnpm/plugin-commands-rebuild@1001.1.5
  - @pnpm/config@1002.2.1

## 1000.0.0

### Major Changes

- 961dc5d: Added a new command for printing the list of dependencies with ignored build scripts: `pnpm ignored-builds` [#8963](https://github.com/pnpm/pnpm/pull/8963).
- 961dc5d: Added a new command for approving dependencies for running scripts during installation: `pnpm approve-builds` [#8963](https://github.com/pnpm/pnpm/pull/8963).

### Patch Changes

- Updated dependencies [f3ffaed]
- Updated dependencies [c96eb2b]
  - @pnpm/config@1002.2.0
  - @pnpm/plugin-commands-rebuild@1001.1.4
  - @pnpm/modules-yaml@1000.1.2
  - @pnpm/read-project-manifest@1000.0.4
