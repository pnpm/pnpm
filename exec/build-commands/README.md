# @pnpm/exec.build-commands

> Commands for managing dependency builds

[![npm version](https://img.shields.io/npm/v/@pnpm/exec.build-commands.svg)](https://www.npmjs.com/package/@pnpm/exec.build-commands)

## Installation

```sh
pnpm add @pnpm/exec.build-commands
```

## License

MIT
