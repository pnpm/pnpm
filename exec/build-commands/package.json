{
  "name": "@pnpm/exec.build-commands",
  "version": "1001.0.19",
  "description": "Commands for managing dependency builds",
  "keywords": [
    "pnpm",
    "pnpm10",
    "rebuild"
  ],
  "license": "MIT",
  "funding": "https://opencollective.com/pnpm",
  "repository": "https://github.com/pnpm/pnpm/blob/main/exec/build-commands",
  "homepage": "https://github.com/pnpm/pnpm/blob/main/exec/build-commands#readme",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js"
  },
  "files": [
    "lib",
    "!*.map"
  ],
  "scripts": {
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "dependencies": {
    "@pnpm/config": "workspace:*",
    "@pnpm/config.config-writer": "workspace:*",
    "@pnpm/modules-yaml": "workspace:*",
    "@pnpm/plugin-commands-rebuild": "workspace:*",
    "@pnpm/prepare-temp-dir": "workspace:*",
    "@pnpm/util.lex-comparator": "catalog:",
    "chalk": "catalog:",
    "enquirer": "catalog:",
    "render-help": "catalog:"
  },
  "peerDependencies": {
    "@pnpm/logger": "catalog:"
  },
  "devDependencies": {
    "@pnpm/exec.build-commands": "workspace:*",
    "@pnpm/plugin-commands-installation": "workspace:*",
    "@pnpm/prepare": "workspace:*",
    "@pnpm/registry-mock": "catalog:",
    "@pnpm/types": "workspace:*",
    "@types/ramda": "catalog:",
    "load-json-file": "catalog:",
    "ramda": "catalog:",
    "read-yaml-file": "catalog:",
    "write-pkg": "catalog:",
    "write-yaml-file": "catalog:"
  },
  "engines": {
    "node": ">=18.12"
  },
  "jest": {
    "preset": "@pnpm/jest-config/with-registry"
  }
}
