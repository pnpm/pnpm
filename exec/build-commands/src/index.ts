import * as approveBuilds from './approveBuilds'
import * as ignoredBuilds from './ignoredBuilds'

export { approveBuilds, ignoredBuilds }
