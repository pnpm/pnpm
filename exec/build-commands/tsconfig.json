{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../__utils__/prepare"
    },
    {
      "path": "../../__utils__/prepare-temp-dir"
    },
    {
      "path": "../../config/config"
    },
    {
      "path": "../../config/config-writer"
    },
    {
      "path": "../../packages/types"
    },
    {
      "path": "../../pkg-manager/modules-yaml"
    },
    {
      "path": "../../pkg-manager/plugin-commands-installation"
    },
    {
      "path": "../plugin-commands-rebuild"
    }
  ]
}
