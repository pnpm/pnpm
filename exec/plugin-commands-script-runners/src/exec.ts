import path from 'path'
import { docsUrl, type RecursiveSummary, throwOnCommandFail, readProjectManifestOnly } from '@pnpm/cli-utils'
import { type Config, types } from '@pnpm/config'
import { makeNodeRequireOption } from '@pnpm/lifecycle'
import { logger } from '@pnpm/logger'
import { tryReadProjectManifest } from '@pnpm/read-project-manifest'
import { sortPackages } from '@pnpm/sort-packages'
import { type Project, type ProjectsGraph } from '@pnpm/types'
import execa from 'execa'
import pLimit from 'p-limit'
import PATH from 'path-name'
import pick from 'ramda/src/pick'
import renderHelp from 'render-help'
import { existsInDir } from './existsInDir'
import { makeEnv } from './makeEnv'
import {
  PARALLEL_OPTION_HELP,
  REPORT_SUMMARY_OPTION_HELP,
  RESUME_FROM_OPTION_HELP,
  shorthands as runShorthands,
} from './run'
import { PnpmError } from '@pnpm/error'
import which from 'which'
import writeJsonFile from 'write-json-file'
import { getNearestProgram, getNearestScript } from './buildCommandNotFoundHint'

export const shorthands: Record<string, string | string[]> = {
  parallel: runShorthands.parallel,
  c: '--shell-mode',
}

export const commandNames = ['exec']

export function rcOptionsTypes (): Record<string, unknown> {
  return {
    ...pick([
      'bail',
      'sort',
      'use-node-version',
      'unsafe-perm',
      'workspace-concurrency',
    ], types),
    'shell-mode': Boolean,
    'resume-from': String,
    'report-summary': Boolean,
  }
}

export const cliOptionsTypes = (): Record<string, unknown> => ({
  ...rcOptionsTypes(),
  recursive: Boolean,
  reverse: Boolean,
})

export function help (): string {
  return renderHelp({
    description: 'Run a shell command in the context of a project.',
    descriptionLists: [
      {
        title: 'Options',

        list: [
          PARALLEL_OPTION_HELP,
          {
            description: 'Run the shell command in every package found in subdirectories \
or every workspace package, when executed inside a workspace. \
For options that may be used with `-r`, see "pnpm help recursive"',
            name: '--recursive',
            shortAlias: '-r',
          },
          {
            description: 'If exist, runs file inside of a shell. \
Uses /bin/sh on UNIX and \\cmd.exe on Windows. \
The shell should understand the -c switch on UNIX or /d /s /c on Windows.',
            name: '--shell-mode',
            shortAlias: '-c',
          },
          RESUME_FROM_OPTION_HELP,
          REPORT_SUMMARY_OPTION_HELP,
        ],
      },
    ],
    url: docsUrl('exec'),
    usages: ['pnpm [-r] [-c] exec <command> [args...]'],
  })
}

export function getResumedPackageChunks ({
  resumeFrom,
  chunks,
  selectedProjectsGraph,
}: {
  resumeFrom: string
  chunks: string[][]
  selectedProjectsGraph: ProjectsGraph
}): string[][] {
  const resumeFromPackagePrefix = Object.keys(selectedProjectsGraph)
    .find((prefix) => selectedProjectsGraph[prefix]?.package.manifest.name === resumeFrom)

  if (!resumeFromPackagePrefix) {
    throw new PnpmError('RESUME_FROM_NOT_FOUND', `Cannot find package ${resumeFrom}. Could not determine where to resume from.`)
  }

  const chunkPosition = chunks.findIndex(chunk => chunk.includes(resumeFromPackagePrefix))
  return chunks.slice(chunkPosition)
}

export async function writeRecursiveSummary (opts: { dir: string, summary: RecursiveSummary }): Promise<void> {
  await writeJsonFile(path.join(opts.dir, 'pnpm-exec-summary.json'), {
    executionStatus: opts.summary,
  })
}

export function createEmptyRecursiveSummary (chunks: string[][]): RecursiveSummary {
  return chunks.flat().reduce<RecursiveSummary>((acc, prefix) => {
    acc[prefix] = { status: 'queued' }
    return acc
  }, {})
}

export function getExecutionDuration (start: [number, number]): number {
  const end = process.hrtime(start)
  return (end[0] * 1e9 + end[1]) / 1e6
}

export async function handler (
  opts: Required<Pick<Config, 'selectedProjectsGraph'>> & {
    bail?: boolean
    unsafePerm?: boolean
    reverse?: boolean
    sort?: boolean
    workspaceConcurrency?: number
    shellMode?: boolean
    resumeFrom?: string
    reportSummary?: boolean
    implicitlyFellbackFromRun?: boolean
  } & Pick<Config, 'extraBinPaths' | 'extraEnv' | 'lockfileDir' | 'modulesDir' | 'dir' | 'userAgent' | 'recursive' | 'workspaceDir' | 'nodeOptions'>,
  params: string[]
): Promise<{ exitCode: number }> {
  // For backward compatibility
  if (params[0] === '--') {
    params.shift()
  }
  const limitRun = pLimit(opts.workspaceConcurrency ?? 4)

  let chunks!: string[][]
  if (opts.recursive) {
    chunks = opts.sort
      ? sortPackages(opts.selectedProjectsGraph)
      : [Object.keys(opts.selectedProjectsGraph).sort()]
    if (opts.reverse) {
      chunks = chunks.reverse()
    }
  } else {
    chunks = [[opts.dir]]
    const project = await tryReadProjectManifest(opts.dir)
    if (project.manifest != null) {
      opts.selectedProjectsGraph = {
        [opts.dir]: {
          dependencies: [],
          package: {
            ...project,
            dir: opts.dir,
          } as Project,
        },
      }
    }
  }

  if (opts.resumeFrom) {
    chunks = getResumedPackageChunks({
      resumeFrom: opts.resumeFrom,
      chunks,
      selectedProjectsGraph: opts.selectedProjectsGraph,
    })
  }

  const result = createEmptyRecursiveSummary(chunks)
  const existsPnp = existsInDir.bind(null, '.pnp.cjs')
  const workspacePnpPath = opts.workspaceDir && await existsPnp(opts.workspaceDir)

  let exitCode = 0
  const prependPaths = [
    './node_modules/.bin',
    ...opts.extraBinPaths,
  ]
  for (const chunk of chunks) {
    // eslint-disable-next-line no-await-in-loop
    await Promise.all(chunk.map(async (prefix: string) =>
      limitRun(async () => {
        result[prefix].status = 'running'
        const startTime = process.hrtime()
        try {
          const pnpPath = workspacePnpPath ?? await existsPnp(prefix)
          const extraEnv = {
            ...opts.extraEnv,
            ...(pnpPath ? makeNodeRequireOption(pnpPath) : {}),
          }
          const env = makeEnv({
            extraEnv: {
              ...extraEnv,
              PNPM_PACKAGE_NAME: opts.selectedProjectsGraph[prefix]?.package.manifest.name,
              ...(opts.nodeOptions ? { NODE_OPTIONS: opts.nodeOptions } : {}),
            },
            prependPaths,
            userAgent: opts.userAgent,
          })
          await execa(params[0], params.slice(1), {
            cwd: prefix,
            env,
            stdio: 'inherit',
            shell: opts.shellMode ?? false,
          })
          result[prefix].status = 'passed'
          result[prefix].duration = getExecutionDuration(startTime)
        } catch (err: any) { // eslint-disable-line
          if (isErrorCommandNotFound(params[0], err, prependPaths)) {
            err.message = `Command "${params[0]}" not found`
            err.hint = await createExecCommandNotFoundHint(params[0], {
              implicitlyFellbackFromRun: opts.implicitlyFellbackFromRun ?? false,
              dir: opts.dir,
              workspaceDir: opts.workspaceDir,
              modulesDir: opts.modulesDir ?? 'node_modules',
            })
          } else if (!opts.recursive && typeof err.exitCode === 'number') {
            exitCode = err.exitCode
            return
          }
          logger.info(err)

          result[prefix] = {
            status: 'failure',
            duration: getExecutionDuration(startTime),
            error: err,
            message: err.message,
            prefix,
          }

          if (!opts.bail) {
            return
          }

          if (!err['code']?.startsWith('ERR_PNPM_')) {
            err['code'] = 'ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL'
          }
          err['prefix'] = prefix
          opts.reportSummary && await writeRecursiveSummary({
            dir: opts.lockfileDir ?? opts.dir,
            summary: result,
          })

          throw err
        }
      }
      )))
  }

  opts.reportSummary && await writeRecursiveSummary({
    dir: opts.lockfileDir ?? opts.dir,
    summary: result,
  })
  throwOnCommandFail('pnpm recursive exec', result)
  return { exitCode }
}

async function createExecCommandNotFoundHint (
  programName: string,
  opts: {
    dir: string
    implicitlyFellbackFromRun: boolean
    workspaceDir?: string
    modulesDir: string
  }
): Promise<string | undefined> {
  if (opts.implicitlyFellbackFromRun) {
    let nearestScript: string | null | undefined
    try {
      nearestScript = getNearestScript(programName, (await readProjectManifestOnly(opts.dir)).scripts)
    } catch (_err) {}
    if (nearestScript) {
      return `Did you mean "pnpm ${nearestScript}"?`
    }
    const nearestProgram = getNearestProgram({
      programName,
      dir: opts.dir,
      workspaceDir: opts.workspaceDir,
      modulesDir: opts.modulesDir,
    })
    if (nearestProgram) {
      return `Did you mean "pnpm ${nearestProgram}"?`
    }
    return undefined
  }
  const nearestProgram = getNearestProgram({
    programName,
    dir: opts.dir,
    workspaceDir: opts.workspaceDir,
    modulesDir: opts.modulesDir,
  })
  if (nearestProgram) {
    return `Did you mean "pnpm exec ${nearestProgram}"?`
  }
  return undefined
}

interface CommandError extends Error {
  originalMessage: string
  shortMessage: string
}

function isErrorCommandNotFound (command: string, error: CommandError, prependPaths: string[]): boolean {
  // Mac/Linux
  if (process.platform === 'linux' || process.platform === 'darwin') {
    return error.originalMessage === `spawn ${command} ENOENT`
  }

  // Windows
  if (process.platform === 'win32') {
    const prepend = prependPaths.join(path.delimiter)
    const whichPath = process.env[PATH] ? `${prepend}${path.delimiter}${process.env[PATH] as string}` : prepend
    return !which.sync(command, {
      nothrow: true,
      path: whichPath,
    })
  }

  return false
}
