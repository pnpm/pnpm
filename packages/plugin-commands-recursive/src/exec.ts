import { WsPkgsGraph } from '@pnpm/config'
import logger from '@pnpm/logger'
import execa = require('execa')
import pLimit from 'p-limit'
import RecursiveSummary from './recursiveSummary'

export default async <T> (
  packageChunks: string[][],
  graph: WsPkgsGraph,
  args: string[],
  cmd: string,
  opts: {
    bail: boolean,
    workspaceConcurrency: number,
    unsafePerm: boolean,
    rawConfig: object,
  },
): Promise<RecursiveSummary> => {
  const limitRun = pLimit(opts.workspaceConcurrency)

  const result = {
    fails: [],
    passes: 0,
  } as RecursiveSummary

  for (const chunk of packageChunks) {
    await Promise.all(chunk.map((prefix: string) =>
      limitRun(async () => {
        try {
          await execa(args[0], args.slice(1), {
            cwd: prefix,
            env: {
              ...process.env,
              PNPM_PACKAGE_NAME: graph[prefix].package.manifest.name,
            },
            stdio: 'inherit',
          })
          result.passes++
        } catch (err) {
          logger.info(err)

          if (!opts.bail) {
            result.fails.push({
              error: err,
              message: err.message,
              prefix,
            })
            return
          }

          // tslint:disable:no-string-literal
          err['code'] = 'ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL'
          err['prefix'] = prefix
          // tslint:enable:no-string-literal
          throw err
        }
      },
    )))
  }

  return result
}
