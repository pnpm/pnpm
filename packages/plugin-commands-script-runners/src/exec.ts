import path from 'path'
import { RecursiveSummary, throwOnCommandFail } from '@pnpm/cli-utils'
import { Config, types } from '@pnpm/config'
import { makeNodeRequireOption } from '@pnpm/lifecycle'
import logger from '@pnpm/logger'
import readProjectManifest from '@pnpm/read-project-manifest'
import sortPackages from '@pnpm/sort-packages'
import execa from 'execa'
import pLimit from 'p-limit'
import PATH from 'path-name'
import pick from 'ramda/src/pick'
import renderHelp from 'render-help'
import existsInDir from './existsInDir'
import {
  PARALLEL_OPTION_HELP,
  shorthands as runShorthands,
} from './run'

export const shorthands = {
  parallel: runShorthands.parallel,
}

export const commandNames = ['exec']

export function rcOptionsTypes () {
  return pick([
    'bail',
    'sort',
    'unsafe-perm',
    'workspace-concurrency',
  ], types)
}

export const cliOptionsTypes = () => ({
  ...rcOptionsTypes(),
  recursive: Boolean,
})

export function help () {
  return renderHelp({
    description: 'Run a shell command in the context of a project.',
    descriptionLists: [
      {
        title: 'Options',

        list: [
          PARALLEL_OPTION_HELP,
          {
            description: 'Run the shell command in every package found in subdirectories \
or every workspace package, when executed inside a workspace. \
For options that may be used with `-r`, see "pnpm help recursive"',
            name: '--recursive',
            shortAlias: '-r',
          },
        ],
      },
    ],
    usages: ['pnpm [-r] exec <command> [args...]'],
  })
}

export async function handler (
  opts: Required<Pick<Config, 'selectedProjectsGraph'>> & {
    bail?: boolean
    unsafePerm?: boolean
    rawConfig: object
    sort?: boolean
    workspaceConcurrency?: number
  } & Pick<Config, 'extraBinPaths' | 'lockfileDir' | 'dir' | 'recursive' | 'workspaceDir'>,
  params: string[]
) {
  // For backward compatibility
  if (params[0] === '--') {
    params.shift()
  }
  const limitRun = pLimit(opts.workspaceConcurrency ?? 4)

  const result = {
    fails: [],
    passes: 0,
  } as RecursiveSummary

  let chunks!: string[][]
  if (opts.recursive) {
    chunks = opts.sort
      ? sortPackages(opts.selectedProjectsGraph)
      : [Object.keys(opts.selectedProjectsGraph).sort()]
  } else {
    chunks = [[opts.dir]]
    opts.selectedProjectsGraph = {
      [opts.dir]: {
        dependencies: [],
        package: {
          ...await readProjectManifest(opts.dir),
          dir: opts.dir,
        },
      },
    }
  }
  const existsPnp = existsInDir.bind(null, '.pnp.cjs')
  const workspacePnpPath = opts.workspaceDir && await existsPnp(opts.workspaceDir)

  for (const chunk of chunks) {
    await Promise.all(chunk.map(async (prefix: string) =>
      limitRun(async () => {
        try {
          const pnpPath = workspacePnpPath ?? await existsPnp(prefix)
          const extraEnv = pnpPath
            ? makeNodeRequireOption(pnpPath)
            : {}
          await execa(params[0], params.slice(1), {
            cwd: prefix,
            env: {
              ...process.env,
              ...extraEnv,
              [PATH]: [
                ...opts.extraBinPaths,
                path.join(opts.dir, 'node_modules/.bin'),
                process.env[PATH],
              ].join(path.delimiter),
              PNPM_PACKAGE_NAME: opts.selectedProjectsGraph[prefix].package.manifest.name,
            },
            stdio: 'inherit',
          })
          result.passes++
        } catch (err) {
          if (!opts.recursive && typeof err.exitCode === 'number') {
            process.exit(err.exitCode)
          }
          logger.info(err)

          if (!opts.bail) {
            result.fails.push({
              error: err,
              message: err.message,
              prefix,
            })
            return
          }

          /* eslint-disable @typescript-eslint/dot-notation */
          err['code'] = 'ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL'
          err['prefix'] = prefix
          /* eslint-enable @typescript-eslint/dot-notation */
          throw err
        }
      }
      )))
  }

  throwOnCommandFail('pnpm recursive exec', result)
}
