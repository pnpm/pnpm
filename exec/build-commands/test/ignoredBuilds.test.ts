import path from 'node:path'
import fs from 'node:fs'
import { ignoredBuilds } from '@pnpm/exec.build-commands'
import { tempDir } from '@pnpm/prepare-temp-dir'
import { writeModulesManifest } from '@pnpm/modules-yaml'

const DEFAULT_MODULES_MANIFEST = {
  hoistedDependencies: {},
  layoutVersion: 4,
  packageManager: '',
  included: {
    optionalDependencies: true,
    dependencies: true,
    devDependencies: true,
  },
  pendingBuilds: [],
  prunedAt: '',
  skipped: [],
  storeDir: '',
  virtualStoreDir: '',
  virtualStoreDirMaxLength: 90,
  registries: {
    default: '',
  },
}

test('ignoredBuilds lists automatically ignored dependencies', async () => {
  const dir = tempDir()
  const modulesDir = path.join(dir, 'node_modules')
  fs.mkdirSync(modulesDir, { recursive: true })
  await writeModulesManifest(modulesDir, {
    ...DEFAULT_MODULES_MANIFEST,
    ignoredBuilds: ['foo'],
  })
  const output = await ignoredBuilds.handler({
    dir,
    modulesDir,
    rootProjectManifest: {},
  })
  expect(output).toMatchSnapshot()
})

test('ignoredBuilds lists explicitly ignored dependencies', async () => {
  const dir = tempDir()
  const modulesDir = path.join(dir, 'node_modules')
  fs.mkdirSync(modulesDir, { recursive: true })
  await writeModulesManifest(modulesDir, {
    ...DEFAULT_MODULES_MANIFEST,
    ignoredBuilds: [],
  })
  const output = await ignoredBuilds.handler({
    dir,
    modulesDir,
    rootProjectManifest: {
      pnpm: {
        ignoredBuiltDependencies: ['bar'],
      },
    },
  })
  expect(output).toMatchSnapshot()
})

test('ignoredBuilds lists both automatically and explicitly ignored dependencies', async () => {
  const dir = tempDir()
  const modulesDir = path.join(dir, 'node_modules')
  fs.mkdirSync(modulesDir, { recursive: true })
  await writeModulesManifest(modulesDir, {
    ...DEFAULT_MODULES_MANIFEST,
    ignoredBuilds: ['foo', 'bar'],
  })
  const output = await ignoredBuilds.handler({
    dir,
    modulesDir,
    rootProjectManifest: {
      pnpm: {
        ignoredBuiltDependencies: ['qar', 'zoo'],
      },
    },
  })
  expect(output).toMatchSnapshot()
})

test('ignoredBuilds prints an info message when there is no node_modules', async () => {
  const dir = tempDir()
  const modulesDir = path.join(dir, 'node_modules')
  const output = await ignoredBuilds.handler({
    dir,
    modulesDir,
    rootProjectManifest: {
      pnpm: {
        ignoredBuiltDependencies: ['qar', 'zoo'],
      },
    },
  })
  expect(output).toMatchSnapshot()
})
