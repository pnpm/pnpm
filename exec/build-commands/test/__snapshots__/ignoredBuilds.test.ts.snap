// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignoredBuilds lists automatically ignored dependencies 1`] = `
"Automatically ignored builds during installation:
  foo
hint: To allow the execution of build scripts for a package, add its name to "pnpm.onlyBuiltDependencies" in your "package.json", then run "pnpm rebuild".
hint: If you don't want to build a package, add it to the "pnpm.ignoredBuiltDependencies" list.
"
`;

exports[`ignoredBuilds lists both automatically and explicitly ignored dependencies 1`] = `
"Automatically ignored builds during installation:
  foo
  bar
hint: To allow the execution of build scripts for a package, add its name to "pnpm.onlyBuiltDependencies" in your "package.json", then run "pnpm rebuild".
hint: If you don't want to build a package, add it to the "pnpm.ignoredBuiltDependencies" list.

Explicitly ignored package builds (via pnpm.ignoredBuiltDependencies):
  qar
  zoo
"
`;

exports[`ignoredBuilds lists explicitly ignored dependencies 1`] = `
"Automatically ignored builds during installation:
  None

Explicitly ignored package builds (via pnpm.ignoredBuiltDependencies):
  bar
"
`;

exports[`ignoredBuilds prints an info message when there is no node_modules 1`] = `
"Automatically ignored builds during installation:
  Cannot identify as no node_modules found

Explicitly ignored package builds (via pnpm.ignoredBuiltDependencies):
  qar
  zoo
"
`;
